### PR TITLE
feat: #247 rp-adapter: evaluate wallet's response against the definitions

### LINKED ISSUE #161: rp adapter: map credentialSubject to rp object
(note: this is a different issue than #127)

The RP adapter is currently sending the contents of the `credentialSubject` directly to the RP. This results in the RP getting an id_token with claims that look like this:

```json
{
    "at_hash": "1MVV-ulQL0Q-gvhQX9Yp_A",
    "aud": [
        "a65f906c-c30c-4e53-81f3-80862bf9644b"
    ],
    "auth_time": 1594239505,
    "exp": 1594243107,
    "iat": 1594239507,
    "id": "did:peer:user",
    "iss": "https://localhost:4444/",
    "jti": "45b00aee-62fa-4a4c-ac01-ca67f6e7adff",
    "nonce": "",
    "rat": 1594239505,
    "sid": "544a57db-fe04-45e0-9b2b-4f14a4cd9c63",
    "stmt": {
        "accountId": "xxxx-xxxx-xxxx-1234",
        "billingPeriod": "P30D",
        "customer": {
            "@type": "Person",
            "name": "Jane Doe"
        },
        "description": "June 2020 Credit Card Statement",
        "minimumPaymentDue": {
            "@type": "PriceSpecification",
            "price": 15,
            "priceCurrency": "CAD"
        },
        "paymentDueDate": "2020-06-30T12:00:00",
        "paymentStatus": "http://schema.org/PaymentDue",
        "totalPaymentDue": {
            "@type": "PriceSpecification",
            "price": 200,
            "priceCurrency": "CAD"
        },
        "url": "http://acmebank.com/invoice.pdf"
    },
    "sub": "26c45ac2-e056-4dd0-acbb-1101fd9a1e7a"
}
```

The RP Adapter should employ rules that define the transformation of one object to the other such that: 

a. the RP understands it, and
b. removes risks of field collisions. Eg. the `id` field in the example above was actually received inside the `credentialSubject`

### LINKED ISSUE #247: rp-adapter: evaluate wallet's response against the definitions
part of #232

### PR DESCRIPTION
**Pull request by llorllale**
closes #247 
closes #161 

----

### Summary

RP Adapter can now request local and remote credentials in any combination (local+remote).

Requests for local and remote credentials are normalized and expressed using standard presentation-exchange criteria in order to evaluate the wallet's response in a uniform and standard manner.

Example:

If a requesting party sends an OIDC request with scope `driver_license:local credit_card_stmt:remote` then the presentation-exchange request delivered to the wallet will look like the following:

<details><summary>Click</summary>

```json
{
    "name": "",
    "purpose": "",
    "input_descriptors": [
        {
            "id": "4eb27312-6676-45ab-8b99-b781413b2414",
            "schema": {
                "name": "Bank Account Information",
                "purpose": "We need your bank and account information.",
                "uri": "https://trustbloc.github.io/context/vc/authorization-credential-v1.jsonld"
            },
            "constraints": {
                "fields": [
                    {
                        "path": ["$.credentialSubject.scope[*].schema.uri"],
                        "filter": {
                            "const": "https://trustbloc.github.io/context/vc/examples-ext-v1.jsonld"
                        }
                    }
                ]
            }
        },
        {
            "id": "3363c23d-fd49-4d98-af61-4d713410d639",
            "schema": {
                "name": "Driver's license.",
                "purpose": "Verify your identity.",
                "uri": "https://trustbloc.github.io/context/vc/examples/mdl-v1.jsonld"
            }
        }                                                                                                                                                                                                          
    ]
}
```

</details>

If everything goes well, the requesting party ends up with an id_token containing the users data like this:

<details><summary>Click</summary>

```json
{
    "at_hash": "JJCNhlQDDaWOcJ5rcz-4aQ",
    "aud": [
        "664864e9-6449-4dda-99cd-bb85508032cf"
    ],
    "auth_time": 1597768660,
    "exp": 1597772262,
    "iat": 1597768662,
    "iss": "https://localhost:4444/",
    "jti": "1f8c47c0-7042-4a52-b89f-dc12e260761d",
    "nonce": "",
    "rat": 1597768660,
    "sid": "ef48b51b-b0c5-4ff9-a8b3-a7b9a600c67f",
    "sub": "40e30c00-10cf-49ad-b8c4-8283f1d9ac74",
    "verified_claims": [                                                                                                                                                                                           
        {
            "document_number": "123-456-789",
            "family_name": "Smith",
            "given_name": "John",
            "id": "did:peer:user"
        },
        {
            "id": "did:peer:user",
            "stmt": {
                "accountId": "xxxx-xxxx-xxxx-1234",
                "billingPeriod": "P30D",
                "customer": {
                    "@type": "Person",
                    "name": "Jane Doe"
                },
                "description": "June 2020 Credit Card Statement",
                "minimumPaymentDue": {
                    "@type": "PriceSpecification",
                    "price": 15,
                    "priceCurrency": "CAD"
                },
                "paymentDueDate": "2020-06-30T12:00:00",
                "paymentStatus": "http://schema.org/PaymentDue",
                "totalPaymentDue": {
                    "@type": "PriceSpecification",
                    "price": 200,
                    "priceCurrency": "CAD"
                },
                "url": "http://acmebank.com/invoice.pdf"
            }
        }
    ]
}
```

</details>

----

### Changes

* feat: remote credentials are expressly asked for in the input descriptors by requesting a top-level schema URI for the authz credential and then constraining the authz scope's value to the final desired credential schema URI
* refactor: PresentationSubmissionJSONLDContext string changed to version hosted by trustbloc since the DIF one is not yet available.
* refactor: #161 all user data is now mapped to a new `verified_claims` claim in the id_token
    
----

Signed-off-by: George Aristy <george.aristy@securekey.com>

### CODE REVIEW

File: test/bdd/fixtures/testdata/presentationdefinitions.json [hunk 1]
```diff
@@ -1,9 +1,63 @@
 {
-  "CreditCardStatement": {
+  "credit_card_stmt:remote": {
```
Comment (sudeshrshetty) [inline test/bdd/fixtures/testdata/presentationdefinitions.json:2]: wrong input descriptor format,

supposed to `"id": "credit_card_stmt:remote",`

Comment (llorllale) [reply test/bdd/fixtures/testdata/presentationdefinitions.json:2]: @sudeshrshetty this new configuration format has been in use since #234

Comment (sudeshrshetty) [reply test/bdd/fixtures/testdata/presentationdefinitions.json:2]: ok, if you aren't sending it as it is to wallet then that's ok

**Comment by llorllale** [comment]
@rolsonquadras @sudeshrshetty 

Notice the JSON path in the constraints for the authz cred:

```json
        {
            "id": "4eb27312-6676-45ab-8b99-b781413b2414",
            "schema": {
                "name": "Bank Account Information",
                "purpose": "We need your bank and account information.",
                "uri": "https://trustbloc.github.io/context/vc/authorization-credential-v1.jsonld"
            },
            "constraints": {
                "fields": [
                    {
                        "path": ["$.credentialSubject.scope[*].schema.uri"],
                        "filter": {
                            "const": "https://trustbloc.github.io/context/vc/examples-ext-v1.jsonld"
                        }
                    }
                ]
            }
        }
```

For this to work we'd need trustbloc/context#49.

File: pkg/restapi/rp/operation/credentials.go [hunk 2]
```diff
@@ -17,76 +17,51 @@ import (
...
 	"github.com/trustbloc/edge-adapter/pkg/vc"
-	"github.com/trustbloc/edge-adapter/pkg/vc/rp"
 )
 
 var errInvalidCredential = errors.New("malformed credential")
 
-func parseWalletResponse(definitions *presexch.PresentationDefinitions, vdriReg vdriapi.Registry,
-	vpBytes []byte) (*vc.AuthorizationCredential, *verifiable.Credential, error) {
+func parseWalletResponse(definitions *presexch.PresentationDefinitions, _ vdriapi.Registry,
+	vpBytes []byte) (local, remote map[string]*verifiable.Credential, err error) {
```
Comment (rolsonquadras) [inline pkg/restapi/rp/operation/credentials.go:25]: Can we remove the unused param `_ vdriapi.Registry` ?

**Comment by llorllale** [comment]
By the way, this claim is not 100% correct:

> RP Adapter can now request local and remote credentials in any combination (local+remote).

As noted in #254 , the `WaitForRemoteCredentials` loop is always being entered, even if there are no remote credentials to fetch. The loop will exit with a timeout error, resulting in a `504 GATEWAY TIMEOUT`.

Fixing #254 should take care of this issue.

File: pkg/restapi/rp/operation/credentials_test.go [hunk 3]
```diff
@@ -17,6 +18,12 @@ import (
 	"testing"
 	"time"
 
+	"github.com/hyperledger/aries-framework-go/pkg/doc/util"
+
+	vc2 "github.com/trustbloc/edge-adapter/pkg/vc"
+
```
Comment (rolsonquadras) [inline pkg/restapi/rp/operation/credentials_test.go:24]: extra space.

File: pkg/restapi/rp/operation/operations.go [hunk 4]
```diff
@@ -744,54 +747,131 @@ func (o *Operation) chapiResponseHandler(w http.ResponseWriter, r *http.Request)
...
 	}
 
-	// TODO Issuer's label on the connection record https://github.com/trustbloc/edge-adapter/issues/93
-	_, err = o.didClient.CreateConnection(crCtx.RPPairwiseDID, issuerDID)
-	if err != nil {
-		msg := fmt.Sprintf("failed to create didcomm connection : %s", err)
-		logger.Errorf(msg)
-		commhttp.WriteErrorResponse(w, http.StatusInternalServerError, msg)
+	// TODO the transient invitationData should be after deleted the CHAPI response is processed but the edge-core
+	//  Store API is missing a Delete() method: https://github.com/trustbloc/edge-core/issues/45
```
Comment (rolsonquadras) [inline pkg/restapi/rp/operation/operations.go:792]: https://github.com/trustbloc/edge-core/issues/45 has been implemented.

Comment (llorllale) [reply pkg/restapi/rp/operation/operations.go:792]: @rolsonquadras Opened #255 and updated TODO

File: pkg/restapi/rp/operation/credentials_test.go (hunk 3 above)
Comment (llorllale) [reply pkg/restapi/rp/operation/credentials_test.go]: @rolsonquadras fixed

File: pkg/restapi/rp/operation/credentials.go (hunk 2 above)
Comment (llorllale) [reply pkg/restapi/rp/operation/credentials.go:25]: @rolsonquadras removed

Comment (llorllale) [reply pkg/restapi/rp/operation/credentials.go:25]: although I have a feeling I'll have to add it again when I test integration with the sandbox... I'll update then (if needed)

**Pull request merged**